File: src/main/java/io/robojackets/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package io.robojackets;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.networktables.BooleanPublisher;
import edu.wpi.first.networktables.BooleanSubscriber;
import edu.wpi.first.networktables.BooleanTopic;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.livewindow.LiveWindow;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import io.robojackets.auto.AutoPrograms;
import io.robojackets.subsystems.Drivetrain;
import io.robojackets.subsystems.RobotMap;

/**
 * The VM is configured to automatically run this class, and to call the functions corresponding to
 * each mode, as described in the TimedRobot documentation. If you change the name of this class or
 * the package after creating this project, you must also update the build.gradle file in the
 * project.
 */
public class Robot extends TimedRobot {
  RobotMap IO = new RobotMap();
  AutoPrograms auton = new AutoPrograms(IO);

  Timer shotTimer = new Timer();
  Timer brakeTimer = new Timer();
  Timer intakeTimer = new Timer();

  Timer seedEncoderTimer = new Timer();

  BooleanTopic calibrationEntry =
      NetworkTableInstance.getDefault().getBooleanTopic("/SmartDashboard/Recalibrate");
  BooleanPublisher calibrationPub = calibrationEntry.publish();
  BooleanSubscriber calibrationSub = calibrationEntry.subscribe(false);

  boolean manualJog;

  /**
   * This function is run when the robot is first started up and should be used for any
   * initialization code.
   */
  @Override
  public void robotInit() {
    LiveWindow.disableAllTelemetry();
    LiveWindow.setEnabled(false);

    IO.ConfigureSystem();

    SmartDashboard.putData("Drivetrain", IO.drivetrain);

    seedEncoderTimer.start();

    IO.drivetrain.ResetYaw(new Rotation2d());
  }

  @Override
  public void robotPeriodic() {
    IO.drivetrain.UpdateOdometry();

    if (!IO.drivetrain.Active() && seedEncoderTimer.get() > 5) {
      System.out.println("Seeding Encoders");

      SmartDashboard.putNumber(
          "Drivetrain/SeedEncoderLastResult", IO.drivetrain.SeedEncoders().value);
      seedEncoderTimer.reset();
    }

    if (IO.mainController.getOptionsButtonPressed()) {
      IO.drivetrain.ResetYaw(new Rotation2d());
    }

    if (IO.mainController.getShareButtonPressed()) {
      IO.drivetrain.SetFieldCentric(!IO.drivetrain.GetFieldCentric());
    }

    for (boolean calib : calibrationSub.readQueueValues()) {
      if (calib) {
        IO.drivetrain.ConfigureSystem();
        calibrationPub.set(false);
      }
    }
  }

  @Override
  public void autonomousInit() {
    auton.Init();
  }

  @Override
  public void autonomousPeriodic() {
    auton.Run();
  }

  @Override
  public void teleopInit() {
    IO.drivetrain.SetFieldCentric(true);
  }

  @Override
  public void teleopPeriodic() {
    double forward =
        deadband(IO.mainController.getLeftY(), 0.1, 1.0)
            * Drivetrain.kMaxSpeedLinearMetersPerSecond;
    double strafe =
        deadband(IO.mainController.getLeftX(), 0.1, 1.0)
            * Drivetrain.kMaxSpeedLinearMetersPerSecond;
    double rotate =
        deadband(IO.mainController.getRightX(), 0.1, 1.0)
            * Drivetrain.kMaxSpeedAngularRadiansPerSecond;

    IO.drivetrain.Drive(forward, strafe, rotate, false);
  }

  @Override
  public void disabledInit() {
    auton.End();
    IO.drivetrain.Stop();
    brakeTimer.reset();
    brakeTimer.start();
  }

  @Override
  public void disabledPeriodic() {}

  @Override
  public void testInit() {}

  @Override
  public void testPeriodic() {}

  @Override
  public void simulationInit() {
    IO.SimInit();
  }

  @Override
  public void simulationPeriodic() {
    IO.SimPeriodic(0, null);
  }

  private double deadband(double val, double min, double max) {
    if (val > max) {
      return max;
    } else if (val < -max) {
      return -max;
    } else if (Math.abs(val) < min) {
      return 0;
    } else {
      double sign = Math.signum(val);

      return sign * (Math.abs(val) - min) / (max - min) * max;
    }
  }
}

File: src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Robot.h"

#include <frc/livewindow/LiveWindow.h>
#include <cmath>                                    // for abs, fabs
#include <exception>                                // for exception
#include <string>                                   // for string
#include "frc/Errors.h"                             // for RuntimeError
#include "lib/PS4Controller.h"                    // for PS4Controller
#include "lib/PneumaticHub.h"                     // for PneumaticHub
#include "units/base.h"                             // for unit_t, operator-
#include "units/math.h"                             // for abs
#include "units/time.h"
#include <iostream>
#include <math.h>
#include <frc/DriverStation.h>
#include <frc/smartdashboard/ListenerExecutor.h>
#include <ntcore_c.h>

using namespace pathplanner;

void Robot::RobotInit()
{
  frc::SmartDashboard::PutNumber("auto/speed (feet per second)", 0);
  frc::SmartDashboard::PutNumber("auto/heading (degrees)", 0);
  frc::SmartDashboard::PutNumber("auto/duration (seconds)", 0);
  // Disable Live Window Stuff, we don't use it...
  frc::LiveWindow::DisableAllTelemetry();
  frc::LiveWindow::SetEnabled(false);

  // System Setup Stuff
  IO.ConfigureSystem();

  frc::SmartDashboard::PutData("Drivetrain", &IO.drivetrain);

  configPub.Set(false);

  // Logging Stuff
  frc::DataLogManager::LogNetworkTables(true);
  IO.RegisterDataEntries(log);
  // arg bool - log joystick data if true
  frc::DriverStation::StartDataLog(log, true);

  seedEncoderTimer.Start();
  IO.drivetrain.ResetYaw(frc::Rotation2d{});

  // localization_flag_entry.SetDefaultBoolean(false);
}

void Robot::RobotPeriodic()
{
  IO.drivetrain.UpdateOdometry();

  IO.UpdateSmartDash();
  autoprograms.SmartDash();

  for (nt::TimestampedBoolean val : configSub.ReadQueue()) {
    if (val.value) {
      IO.drivetrain.ConfigureSystem();
    }
  }

  // Logging Stuff
  IO.LogDataEntries(log);

  if (!IO.drivetrain.Active() && seedEncoderTimer.Get() > 5_s) {
      std::cout << "seeding encoders" << std::endl;

      zeroPub.Set(IO.drivetrain.SeedEncoders());
      seedEncoderTimer.Reset();
  }

  if (IO.mainController.GetOptionsButtonPressed())
  {
    IO.drivetrain.ResetYaw(0_deg);
  }

  if (IO.mainController.GetShareButtonPressed())
  {
    IO.drivetrain.SetFieldCentric(!IO.drivetrain.GetFieldCentric());
  }
}

void Robot::AutonomousInit()
{
  autoprograms.Init();
}

void Robot::AutonomousPeriodic()
{
  autoprograms.Run();
}

void Robot::TeleopInit()
{
  IO.drivetrain.SetFieldCentric(true);
}

void Robot::TeleopPeriodic()
{
  // DRIVE CODE
  auto forward = deadband(IO.mainController.GetLeftY(), 0.1, 1.0) * Drivetrain::kMaxSpeedLinear;
  auto strafe = deadband(IO.mainController.GetLeftX(), 0.1, 1.0) * Drivetrain::kMaxSpeedLinear;
  auto rotate = deadband(IO.mainController.GetRightX(), 0.1, 1.0) * Drivetrain::kMaxSpeedAngular;

  // std::cout << (forward / 1_mps).value() << ", " << (strafe / 1_mps).value() << ", " << (rotate / 1_rad_per_s).value() << std::endl;

  IO.drivetrain.Drive(frc::ChassisSpeeds{forward, strafe, rotate}, false);
}

void Robot::DisabledInit()
{
  IO.drivetrain.Stop();
}

void Robot::DisabledPeriodic()
{

}

void Robot::SimulationInit() {
  IO.SimInit();
}

void Robot::SimulationPeriodic()
{
  IO.SimPeriodic();
}

void Robot::TestInit()
{
  IO.pdp.ClearStickyFaults();
}

void Robot::TestPeriodic()
{

}

double Robot::deadband(double val, double min, double max)
{
  if (val > max)
  {
    return max;
  }
  else if (val < -max)
  {
    return -max;
  }
  else if (std::abs(val) < min)
  {
    return 0.0;
  }
  else
  {
    double sgn = val / std::abs(val);
    // return sgn * (std::abs(val) - min) / (max - min) * max;
    return sgn * (std::abs(val) - min) / (max - min) * max;
  }
}

#ifndef RUNNING_FRC_TESTS
int main()
{
  return frc::StartRobot<Robot>();
}
#endif
